### DISCUSSION TITLE
Type error on useMutation - No overload matches this call.

### DISCUSSION THREAD

**Discussion by canrozanes** [TanStack/query#4658] [Q&A]
Working on simple Password reset form and I've got the following:

```ts

class API {
   updatePassword(body: UpdatePasswordRequestBody): Promise<UpdatePasswordResponse>
}

// Inside the component
  const ChangePasswordForm = () => {
  const updatePasswordMutation = useMutation<UpdatePasswordResponse, string>({
      mutationFn: API.updatePassword,
    });

  return  (
     <>some jsx</>
  )
}
```

For the mutationFn line, I get the following type error:

```
No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ mutationFn: (values: UpdatePasswordResponse) => Promise<UpdatePasswordResponse>; onSuccess: () => void; }' is not assignable to parameter of type 'MutationKey'.
      Object literal may only specify known properties, and 'mutationFn' does not exist in type 'readonly unknown[]'.
```

Any ideas on how I could get around this?

**Comment by TkDodo** [comment]
just do:

```
  const updatePasswordMutation = useMutation({
      mutationFn: API.updatePassword,
    });
```

[working TypeScript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCyyYEMbAQB2cAvnAGZQQhwDkAAnkangMYDWA9FOrmzAC0AR2TooATzoAoaTAlh0cAKpgAJvnQAFXKlQB3aGoBK6UelYAhCGolwAvIkoQIALjisowIgHMychSVVDRhtXQMjU1RIFiVHJAAjXCh3ImQQBPF-aTZiVjgAQS0ASQc4RGlylHVNHT1DKDUACgSbCXdg2vCGkzMxKzaASnctahBgDAAeTtC6iMaomIwAPgRKqt4YZCgSUZoJ9AA6XlQIABsAN3QmxOT3AEZSQfX-UlkuLjhilmA1JRgABZKXLgYjoIgwda5FjwADCANwvjC9SMADFoLRHE1Bg5lhVytD8sgarNukZsHgCMQymgsDh8IQiDcXlUQPSqURUUR3EViodiSFkfM1AAaFlPADcsnKm22JAARKcQEoAFaoAAe8tkpCAA)

**Reply by canrozanes** [reply]
Thanks @TkDodo.  That does indeed work and the request/response object types are inferred properly . However, then my `updatePasswordMutation.error` is `unknown`. Is there anyway I can help TS infer that as well? 

[Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCyyYEMbAQB2cAvnAGZQQhwDkAAnkangMYDWA9FOrmzAC0AR2TooATzoAoUJFhwASnwEAaREpUwAchAAm6MpWq06vfjBnSYEsIYCqYPfnQAFXKlQB3aHuWj0VgAhfQk4AF4NCggIAC44VihgIgBzMmtbBycXd08fKD9AyBZDSKQAI1woeKJkEHLxdOk2YlY4AEFXAEkIuERpPpRsmDcPb18ACnLQ+MdnEdzxgv8xYNCASnjXE2AMAB45nLH8wtRijAA+BAHB3hhkKBJtml30ADpeVAgAGwA3dAmFSq8QAjKR1jd0qRpNIuFw4F0WMADHAYAALQwtcDEdBEGDNVrwADCaNwqVGeV8ADFoLRIhN1hELv0+i0WPBkMMKUs9Ng8ARiL00FgcPhCERAZC4CBRQKiFSiPFOl03pz5tyTqobuCANwwvrACgTNVHSkFPli4hvcTUKDra6DOB3B4kPaYQKoXApdBXE0LY6+C1y61QW2kPZcd2eL0+uBwhFwFrIb56RMeeDAeAwCBwAAGHgSMCSqVzcHKOFRpMsqDguDLmZQKW+Em1MOdjzgACIviBDAArVAAD07MOhGTscCjnu9zzANbKNzYaOAKd4is0Fl0Bi1Y7ZbSnMd6EzA1Dn8QPM9PqEZ4WZDqd6HuHb2ZzJVxPEDnbyXK70a-DXCvkQFzSKQQA) is a working Typescript playground of the issue.

**Reply by TkDodo** [reply]
this is expected as we cannot know what your function is rejecting the Promise with. Promises can only have a type for the fulfilled state. I've written about that here:

https://tkdodo.eu/blog/react-query-and-type-script#what-about-error

Your example shows this quite well, too:

```
if(updatePasswordMutation.error){
  return <Message>{updatePasswordMutation.error}</Message> // I could cast it to `as string` but that's a bit ugly
}
```

why would `error` be of type string? If it's an instanceOf Error, it should be `.error.message`. Have a look at [this playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCyyYEMbAQB2cAvnAGZQQhwDkAAnkangMYDWA9FOrmzAC0AR2TooATzoAoUJFhwASnwEAaREpUwAchAAm6MpWq06vfjBnSYEsIYCqYPfnQAFXKlQB3aHuWj0VgAhfQk4AF4NCggIAC44VihgIgBzMmtbBycXd08fKD9AyBZDSKQAI1woeKJkEHLxdOk2YlY4AEFXAEkIuERpPpRsmDcPb18ACnLQ+MdnEdzxgv8xYNCASnjXE2AMAB45nLH8wtRijAA+BAHB3hhkKBJtml30ADpeVAgAGwA3dAmFSq8QAjKR1jd0qRpNIuFw4F0WMADHAYAALQwtcDEdBEGDNVrwADCaNwqVGeV8ADFoLRIhN1hELv0+i0WPBkMMKUs9Ng8ARiL00FgcPhCERAZC4CBRQKiFSiPFOl03pz5tyTqobuCANwwvrACgTNVHSkFPli4hvcTUKBwZKsMlsdAQChwACiUFt62ugzgdweJD2mECqFwKXQVxNC2OvgtcutXugbxAofD6FIey4Ic86eZcIRcBayG+eiLHngwHgMAgcAABh4EjAkqk63ByjhUaTLKg4Lh21WUClvhJtTCA484AAiL6puAAK1QAA8pzDoRk7HAc2GI88wL2yjc2GjgKXeIrNBZdAYteu2W1t+nehMwNR9-FH7u36hGeFmb7-XQe5Jz2M4ySuV8IH3N5j1PPRz0zLgwKIC5pFIIA).

That said, we do plan on changing the type of error to default to `Error` in v5:
- https://github.com/TanStack/query/issues/4676
